File: apps/extractor/typescript/src/extractor.ts
import {
  DocSet,
  DocSetUtils,
  Project,
  SupportedLanguages,
  DocItem,
  ItemKind,
  DocBlock,
  DocTag,
} from '@opendocs/model';
import * as ts from 'typescript';
import * as path from 'path';
import * as fs from 'fs';

export interface ExtractorOptions {
  tsconfigPath: string;
  projectName?: string;
  projectId?: string;
  repoUrl?: string;
  repoType?: string;
  fileUrlTemplate?: string;
}

/**
 * Context for tracking FQN generation during extraction
 */
interface ExtractionContext {
  packageName: string;      // From package.json
  projectRoot: string;       // Project root directory
  parentIds: string[];       // Stack of parent FQNs for nested items
}

/**
 * Build a fully qualified name for a DocItem
 * Format: package#Symbol or package#Parent#Child
 */
function buildFQN(context: ExtractionContext, name: string): string {
  const { packageName, parentIds } = context;

  if (parentIds.length === 0) {
    return `${packageName}#${name}`;  // Top-level: package#Symbol
  }

  const parentFQN = parentIds[parentIds.length - 1];
  return `${parentFQN}#${name}`;  // Nested: parent#child
}

/**
 * Extract OpenDocs documentation from a TypeScript project
 */
export async function extractDocumentation(options: ExtractorOptions): Promise<DocSet> {
  const { tsconfigPath, projectName, projectId, repoUrl, repoType, fileUrlTemplate } = options;
  const projectRoot = path.dirname(tsconfigPath);
  const packageName = getPackageName(projectRoot);

  // Read and parse tsconfig
  const configFile = ts.readConfigFile(tsconfigPath, ts.sys.readFile);
  if (configFile.error) {
    throw new Error(`Error reading tsconfig: ${configFile.error.messageText}`);
  }

  const parsedConfig = ts.parseJsonConfigFileContent(
    configFile.config,
    ts.sys,
    projectRoot
  );

  // Create TypeScript program
  const program = ts.createProgram(parsedConfig.fileNames, parsedConfig.options);
  const checker = program.getTypeChecker();

  // Create DocSet
  const docSet = DocSetUtils.create({
    id: projectId || projectName || path.basename(projectRoot),
    name: projectName || path.basename(projectRoot),
    generator: {
      name: '@opendocs/extractor-typescript',
      version: '0.2.0',
    },
  });

  // Create Project
  const project: Project = {
    id: projectId || projectName || path.basename(projectRoot),
    name: projectName || path.basename(projectRoot),
    language: SupportedLanguages.TYPESCRIPT,
    version: getPackageVersion(projectRoot),
    items: [],
  };

  // Add repository info if provided
  if (repoUrl) {
    project.repository = {
      type: repoType || 'git',
      url: repoUrl,
      ...(fileUrlTemplate && { fileUrlTemplate }),
    };
  }

  // Create extraction context
  const context: ExtractionContext = {
    packageName,
    projectRoot,
    parentIds: [],
  };

  // Extract documentation from source files
  for (const sourceFile of program.getSourceFiles()) {
    if (!sourceFile.isDeclarationFile && !sourceFile.fileName.includes('node_modules')) {
      const items = extractFromSourceFile(sourceFile, checker, context);
      project.items?.push(...items);
    }
  }

  docSet.projects.push(project);

  return docSet;
}

/**
 * Extract documentation items from a source file
 */
function extractFromSourceFile(sourceFile: ts.SourceFile, checker: ts.TypeChecker, context: ExtractionContext): DocItem[] {
  const items: DocItem[] = [];

  ts.forEachChild(sourceFile, (node) => {
    const item = extractFromNode(node, sourceFile, checker, context);
    if (item) {
      items.push(item);
    }
  });

  return items;
}

/**
 * Extract a DocItem from a TypeScript node
 */
function extractFromNode(
  node: ts.Node,
  sourceFile: ts.SourceFile,
  checker: ts.TypeChecker,
  context: ExtractionContext
): DocItem | null {
  // Only process exported declarations
  if (!isNodeExported(node)) {
    return null;
  }

  let item: DocItem | null = null;

  if (ts.isClassDeclaration(node) && node.name) {
    item = extractClass(node, sourceFile, checker, context);
  } else if (ts.isFunctionDeclaration(node) && node.name) {
    item = extractFunction(node, sourceFile, checker, context);
  } else if (ts.isInterfaceDeclaration(node)) {
    item = extractInterface(node, sourceFile, checker, context);
  } else if (ts.isEnumDeclaration(node)) {
    item = extractEnum(node, sourceFile, checker, context);
  } else if (ts.isTypeAliasDeclaration(node)) {
    item = extractTypeAlias(node, sourceFile, checker, context);
  }

  return item;
}

/**
 * Extract a class declaration
 */
function extractClass(
  node: ts.ClassDeclaration,
  sourceFile: ts.SourceFile,
  checker: ts.TypeChecker,
  context: ExtractionContext
): DocItem {
  const name = node.name?.text || 'AnonymousClass';
  const fqn = buildFQN(context, name);

  // Create a new context for nested items
  const nestedContext: ExtractionContext = {
    ...context,
    parentIds: [...context.parentIds, fqn],
  };

  const item: DocItem = {
    id: fqn,
    name,
    kind: ItemKind.CLASS,
    language: SupportedLanguages.TYPESCRIPT,
    location: getLocation(node, sourceFile, context.projectRoot),
    docBlock: extractDocBlock(node),
    children: [],
  };

  // Extract members
  node.members.forEach((member) => {
    const memberItem = extractClassMember(member, sourceFile, checker, nestedContext);
    if (memberItem) {
      memberItem.parentId = fqn;
      item.children?.push(memberItem);
    }
  });

  return item;
}

/**
 * Extract a class member (method, property, etc.)
 */
function extractClassMember(
  member: ts.ClassElement,
  sourceFile: ts.SourceFile,
  checker: ts.TypeChecker,
  context: ExtractionContext
): DocItem | null {
  if (ts.isMethodDeclaration(member) && member.name && ts.isIdentifier(member.name)) {
    return extractMethod(member, sourceFile, checker, context);
  } else if (ts.isPropertyDeclaration(member) && member.name && ts.isIdentifier(member.name)) {
    return extractProperty(member, sourceFile, checker, context);
  } else if (ts.isConstructorDeclaration(member)) {
    return extractConstructor(member, sourceFile, checker, context);
  }

  return null;
}

/**
 * Extract a method declaration
 */
function extractMethod(
  node: ts.MethodDeclaration,
  sourceFile: ts.SourceFile,
  checker: ts.TypeChecker,
  context: ExtractionContext
): DocItem {
  const name = (node.name as ts.Identifier).text;
  const fqn = buildFQN(context, name);

  return {
    id: fqn,
    name,
    kind: ItemKind.METHOD,
    language: SupportedLanguages.TYPESCRIPT,
    location: getLocation(node, sourceFile, context.projectRoot),
    docBlock: extractDocBlock(node),
    metadata: {
      visibility: getVisibility(node),
      isStatic: hasModifier(node, ts.SyntaxKind.StaticKeyword),
      signature: {
        parameters: extractParameters(node, checker),
        returnType: extractReturnType(node, checker),
      },
    },
  };
}

/**
 * Extract a property declaration
 */
function extractProperty(
  node: ts.PropertyDeclaration,
  sourceFile: ts.SourceFile,
  checker: ts.TypeChecker,
  context: ExtractionContext
): DocItem {
  const name = (node.name as ts.Identifier).text;
  const fqn = buildFQN(context, name);

  return {
    id: fqn,
    name,
    kind: ItemKind.PROPERTY,
    language: SupportedLanguages.TYPESCRIPT,
    location: getLocation(node, sourceFile, context.projectRoot),
    docBlock: extractDocBlock(node),
    metadata: {
      visibility: getVisibility(node),
      isStatic: hasModifier(node, ts.SyntaxKind.StaticKeyword),
      isReadonly: hasModifier(node, ts.SyntaxKind.ReadonlyKeyword),
      type: node.type ? { name: node.type.getText() } : undefined,
    },
  };
}

/**
 * Extract a constructor declaration
 */
function extractConstructor(
  node: ts.ConstructorDeclaration,
  sourceFile: ts.SourceFile,
  checker: ts.TypeChecker,
  context: ExtractionContext
): DocItem {
  const fqn = buildFQN(context, 'constructor');

  return {
    id: fqn,
    name: 'constructor',
    kind: ItemKind.CONSTRUCTOR,
    language: SupportedLanguages.TYPESCRIPT,
    location: getLocation(node, sourceFile, context.projectRoot),
    docBlock: extractDocBlock(node),
    metadata: {
      visibility: getVisibility(node),
      signature: {
        parameters: extractParameters(node, checker),
      },
    },
  };
}

/**
 * Extract a function declaration
 */
function extractFunction(
  node: ts.FunctionDeclaration,
  sourceFile: ts.SourceFile,
  checker: ts.TypeChecker,
  context: ExtractionContext
): DocItem {
  const name = node.name?.text || 'AnonymousFunction';
  const fqn = buildFQN(context, name);

  return {
    id: fqn,
    name,
    kind: ItemKind.FUNCTION,
    language: SupportedLanguages.TYPESCRIPT,
    location: getLocation(node, sourceFile, context.projectRoot),
    docBlock: extractDocBlock(node),
    metadata: {
      signature: {
        parameters: extractParameters(node, checker),
        returnType: extractReturnType(node, checker),
      },
    },
  };
}

/**
 * Extract an interface declaration
 */
function extractInterface(
  node: ts.InterfaceDeclaration,
  sourceFile: ts.SourceFile,
  checker: ts.TypeChecker,
  context: ExtractionContext
): DocItem {
  const name = node.name.text;
  const fqn = buildFQN(context, name);

  // Create a new context for nested items
  const nestedContext: ExtractionContext = {
    ...context,
    parentIds: [...context.parentIds, fqn],
  };

  const item: DocItem = {
    id: fqn,
    name,
    kind: ItemKind.INTERFACE,
    language: SupportedLanguages.TYPESCRIPT,
    location: getLocation(node, sourceFile, context.projectRoot),
    docBlock: extractDocBlock(node),
    children: [],
  };

  // Extract members
  node.members.forEach((member) => {
    if (ts.isPropertySignature(member) && member.name && ts.isIdentifier(member.name)) {
      const memberFQN = buildFQN(nestedContext, member.name.text);
      item.children?.push({
        id: memberFQN,
        name: member.name.text,
        kind: ItemKind.PROPERTY,
        language: SupportedLanguages.TYPESCRIPT,
        location: getLocation(member, sourceFile, context.projectRoot),
        parentId: fqn,
        docBlock: extractDocBlock(member),
        metadata: {
          type: member.type ? { name: member.type.getText() } : undefined,
        },
      });
    }
  });

  return item;
}

/**
 * Extract an enum declaration
 */
function extractEnum(
  node: ts.EnumDeclaration,
  sourceFile: ts.SourceFile,
  checker: ts.TypeChecker,
  context: ExtractionContext
): DocItem {
  const name = node.name.text;
  const fqn = buildFQN(context, name);

  // Create a new context for nested items
  const nestedContext: ExtractionContext = {
    ...context,
    parentIds: [...context.parentIds, fqn],
  };

  const item: DocItem = {
    id: fqn,
    name,
    kind: ItemKind.ENUM,
    language: SupportedLanguages.TYPESCRIPT,
    location: getLocation(node, sourceFile, context.projectRoot),
    docBlock: extractDocBlock(node),
    children: [],
  };

  // Extract enum members
  node.members.forEach((member) => {
    if (ts.isIdentifier(member.name)) {
      const memberFQN = buildFQN(nestedContext, member.name.text);
      item.children?.push({
        id: memberFQN,
        name: member.name.text,
        kind: ItemKind.ENUM_MEMBER,
        language: SupportedLanguages.TYPESCRIPT,
        location: getLocation(member, sourceFile, context.projectRoot),
        parentId: fqn,
        docBlock: extractDocBlock(member),
      });
    }
  });

  return item;
}

/**
 * Extract a type alias declaration
 */
function extractTypeAlias(
  node: ts.TypeAliasDeclaration,
  sourceFile: ts.SourceFile,
  checker: ts.TypeChecker,
  context: ExtractionContext
): DocItem {
  const name = node.name.text;
  const fqn = buildFQN(context, name);

  return {
    id: fqn,
    name,
    kind: ItemKind.TYPE_ALIAS,
    language: SupportedLanguages.TYPESCRIPT,
    location: getLocation(node, sourceFile, context.projectRoot),
    docBlock: extractDocBlock(node),
    metadata: {
      type: { name: node.type.getText() },
    },
  };
}

/**
 * Extract JSDoc comments into a DocBlock
 */
function extractDocBlock(node: ts.Node): DocBlock | undefined {
  const jsDocTags = ts.getJSDocTags(node);
  const jsDocComments = ts.getJSDocCommentsAndTags(node);

  if (jsDocComments.length === 0) {
    return undefined;
  }

  const docBlock: DocBlock = {};

  // Extract content from JSDoc comment
  for (const comment of jsDocComments) {
    if (ts.isJSDoc(comment) && comment.comment) {
      const commentText =
        typeof comment.comment === 'string'
          ? comment.comment
          : comment.comment.map((c) => c.text).join('');
      docBlock.content = commentText;
      break;
    }
  }

  // Tags - use Record<string, (string | DocTag)[]> format per spec
  const tags: Record<string, any[]> = {};

  // Extract tags
  for (const tag of jsDocTags) {
    const tagName = tag.tagName.text;
    const commentText =
      typeof tag.comment === 'string'
        ? tag.comment
        : tag.comment?.map((c) => c.text).join('') || '';

    // Handle @param tags specially
    if (ts.isJSDocParameterTag(tag) && tag.name && ts.isIdentifier(tag.name)) {
      const docTag: DocTag = {
        name: tagName,
        content: commentText,
        parameters: {
          name: tag.name.text,
        },
      };

      if (tag.typeExpression) {
        docTag.parameters!.type = tag.typeExpression.getText();
      }

      if (!tags[tagName]) {
        tags[tagName] = [];
      }
      tags[tagName].push(docTag);
    } else if (tagName === 'returns' || tagName === 'return') {
      // Simple string for returns per spec examples
      if (!tags.returns) {
        tags.returns = [];
      }
      tags.returns.push(commentText);
    } else {
      // Other tags
      const docTag: DocTag = {
        name: tagName,
        content: commentText,
      };

      if (!tags[tagName]) {
        tags[tagName] = [];
      }
      tags[tagName].push(docTag);
    }
  }

  if (Object.keys(tags).length > 0) {
    docBlock.tags = tags;
  }

  return docBlock.content || docBlock.tags ? docBlock : undefined;
}

/**
 * Extract parameters from a function-like declaration
 */
function extractParameters(
  node: ts.FunctionLikeDeclaration,
  checker: ts.TypeChecker
): any[] {
  return node.parameters.map((param) => ({
    name: param.name.getText(),
    type: param.type ? { name: param.type.getText() } : undefined,
    isOptional: !!param.questionToken,
    isRest: !!param.dotDotDotToken,
  }));
}

/**
 * Extract return type from a function-like declaration
 */
function extractReturnType(node: ts.FunctionLikeDeclaration, checker: ts.TypeChecker): any {
  if (node.type) {
    return { name: node.type.getText() };
  }
  return undefined;
}

/**
 * Get source location for a node
 */
function getLocation(
  node: ts.Node,
  sourceFile: ts.SourceFile,
  projectRoot: string
): { path: string; number: number; column: number } {
  const { line, character } = sourceFile.getLineAndCharacterOfPosition(node.getStart());
  return {
    path: path.relative(projectRoot, sourceFile.fileName),
    number: line + 1,
    column: character + 1,
  };
}

/**
 * Get visibility modifier
 */
function getVisibility(node: ts.Node): 'public' | 'private' | 'protected' | undefined {
  if (hasModifier(node, ts.SyntaxKind.PrivateKeyword)) return 'private';
  if (hasModifier(node, ts.SyntaxKind.ProtectedKeyword)) return 'protected';
  if (hasModifier(node, ts.SyntaxKind.PublicKeyword)) return 'public';
  return 'public'; // Default in TypeScript
}

/**
 * Check if a node has a specific modifier
 */
function hasModifier(node: ts.Node, kind: ts.SyntaxKind): boolean {
  return (
    ts.canHaveModifiers(node) &&
    !!ts.getModifiers(node)?.some((mod) => mod.kind === kind)
  );
}

/**
 * Check if a node is exported
 */
function isNodeExported(node: ts.Node): boolean {
  return hasModifier(node, ts.SyntaxKind.ExportKeyword) || hasModifier(node, ts.SyntaxKind.DefaultKeyword);
}

/**
 * Get package version from package.json
 */
function getPackageVersion(dir: string): string | undefined {
  const packageJsonPath = path.join(dir, 'package.json');
  if (fs.existsSync(packageJsonPath)) {
    try {
      const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf-8'));
      return packageJson.version;
    } catch {
      return undefined;
    }
  }
  return undefined;
}

/**
 * Get package name from package.json for FQN generation
 */
function getPackageName(dir: string): string {
  const packageJsonPath = path.join(dir, 'package.json');
  if (fs.existsSync(packageJsonPath)) {
    try {
      const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf-8'));
      return packageJson.name || path.basename(dir);
    } catch {
      return path.basename(dir);
    }
  }
  return path.basename(dir);
}

File: apps/extractor/python/opendocs_extractor/extractor.py
"""Python documentation extractor using AST."""

import ast
import sys
from datetime import datetime
from pathlib import Path
from typing import List, Optional, Dict, Any
from dataclasses import dataclass

from docstring_parser import parse as parse_docstring
from opendocs_model import (
    DocSet,
    DocItem,
    DocBlock,
    DocTag,
    Parameter,
    Location,
    Language,
    ItemKind,
    TagName,
    VERSION,
    Repository,
)


@dataclass
class ExtractionContext:
    """Context for tracking FQN generation during extraction."""
    module_path: str              # e.g., "calculator"
    current_class: Optional[str] = None  # e.g., "Calculator"
    parent_ids: List[str] = None  # Stack of parent FQNs

    def __post_init__(self):
        if self.parent_ids is None:
            self.parent_ids = []


def get_module_path(file_path: Path, source_dir: Path) -> str:
    """Convert file path to module path (calculator.py → calculator)."""
    rel_path = file_path.relative_to(source_dir)

    if rel_path.name == "__init__.py":
        return str(rel_path.parent).replace("/", ".")

    return str(rel_path.with_suffix("")).replace("/", ".")


def build_fqn(module_path: str, class_name: Optional[str], item_name: str) -> str:
    """Build Python FQN: module.Class.method"""
    parts = [module_path]
    if class_name:
        parts.append(class_name)
    parts.append(item_name)
    return ".".join(parts)


def extract_documentation(
    source_dir: Path,
    project_name: Optional[str] = None,
    project_id: Optional[str] = None,
    project_version: Optional[str] = None,
    repo_url: Optional[str] = None,
    repo_type: Optional[str] = None,
    file_url_template: Optional[str] = None,
) -> DocSet:
    """
    Extract OpenDocs documentation from a Python project.

    Args:
        source_dir: Directory containing Python source files
        project_name: Optional project name
        project_id: Optional project ID
        project_version: Optional project version

    Returns:
        DocSet representing the OpenDocs documentation
    """
    # Create DocSet
    doc_set: DocSet = {
        "id": project_id or project_name or source_dir.name,
        "name": project_name or source_dir.name,
        "version": VERSION,
        "format": "json",
        "projects": [],
        "metadata": {
            "created": datetime.now().isoformat(),
            "modified": datetime.now().isoformat(),
            "generator": {
                "name": "opendocs-extractor-python",
                "version": "0.2.0",
            },
        },
    }

    # Create Project
    project: DocSet["projects"][0] = {  # type: ignore
        "id": project_id or project_name or source_dir.name,
        "name": project_name or source_dir.name,
        "language": Language.PYTHON,
        "version": project_version or get_project_version(source_dir),
        "items": [],
    }

    # Add repository info if provided
    if repo_url:
        project["repository"] = Repository(
            type=repo_type or "git",
            url=repo_url,
            fileUrlTemplate=file_url_template,
        )

    # Extract from Python files
    for py_file in source_dir.rglob("*.py"):
        if should_process_file(py_file, source_dir):
            items = extract_from_file(py_file, source_dir)
            project["items"].extend(items)

    doc_set["projects"].append(project)

    return doc_set


def should_process_file(file_path: Path, source_dir: Path) -> bool:
    """Check if a file should be processed."""
    # Skip test files, __pycache__, etc.
    parts = file_path.relative_to(source_dir).parts
    skip_patterns = ["test_", "__pycache__", ".venv", "venv", "build", "dist"]

    for part in parts:
        if any(pattern in str(part) for pattern in skip_patterns):
            return False

    return True


def extract_from_file(file_path: Path, source_dir: Path) -> List[DocItem]:
    """Extract documentation items from a Python file."""
    try:
        with file_path.open("r", encoding="utf-8") as f:
            source = f.read()

        tree = ast.parse(source, filename=str(file_path))
        items: List[DocItem] = []

        # Create extraction context for this file
        module_path = get_module_path(file_path, source_dir)
        context = ExtractionContext(module_path=module_path)

        for node in ast.iter_child_nodes(tree):
            item = extract_from_node(node, file_path, source_dir, context)
            if item:
                items.append(item)

        return items

    except Exception as e:
        print(f"Warning: Failed to parse {file_path}: {e}", file=sys.stderr)
        return []


def extract_from_node(
    node: ast.AST, file_path: Path, source_dir: Path, context: ExtractionContext
) -> Optional[DocItem]:
    """Extract a DocItem from an AST node."""
    if isinstance(node, ast.ClassDef):
        return extract_class(node, file_path, source_dir, context)
    elif isinstance(node, ast.FunctionDef) or isinstance(node, ast.AsyncFunctionDef):
        # Only extract module-level functions
        return extract_function(node, file_path, source_dir, context)

    return None


def extract_class(node: ast.ClassDef, file_path: Path, source_dir: Path, context: ExtractionContext) -> DocItem:
    """Extract a class declaration."""
    fqn = build_fqn(context.module_path, None, node.name)

    # Create a new context for nested items
    nested_context = ExtractionContext(
        module_path=context.module_path,
        current_class=node.name,
        parent_ids=context.parent_ids + [fqn]
    )

    item: DocItem = {
        "id": fqn,
        "name": node.name,
        "kind": ItemKind.CLASS,
        "language": Language.PYTHON,
        "location": get_location(node, file_path, source_dir),
        "children": [],
    }

    # Extract docstring
    docblock = extract_docblock(node)
    if docblock:
        item["docBlock"] = docblock

    # Extract methods and properties
    for child in node.body:
        if isinstance(child, (ast.FunctionDef, ast.AsyncFunctionDef)):
            method = extract_method(child, file_path, source_dir, nested_context)
            if method:
                method["parent_id"] = fqn
                if "children" not in item:
                    item["children"] = []
                item["children"].append(method)

    return item


def extract_method(
    node: ast.FunctionDef | ast.AsyncFunctionDef, file_path: Path, source_dir: Path, context: ExtractionContext
) -> DocItem:
    """Extract a method declaration."""
    # Determine visibility based on naming convention
    visibility = "private" if node.name.startswith("_") else "public"

    fqn = build_fqn(context.module_path, context.current_class, node.name)

    item: DocItem = {
        "id": fqn,
        "name": node.name,
        "kind": ItemKind.CONSTRUCTOR if node.name == "__init__" else ItemKind.METHOD,
        "language": Language.PYTHON,
        "location": get_location(node, file_path, source_dir),
        "metadata": {
            "visibility": visibility,
            "signature": {
                "parameters": extract_parameters(node),
            },
        },
    }

    # Extract docstring
    docblock = extract_docblock(node)
    if docblock:
        item["docBlock"] = docblock

    # Extract return type annotation
    if node.returns:
        item["metadata"]["signature"]["returnType"] = {"name": ast.unparse(node.returns)}

    return item


def extract_function(
    node: ast.FunctionDef | ast.AsyncFunctionDef, file_path: Path, source_dir: Path, context: ExtractionContext
) -> DocItem:
    """Extract a function declaration."""
    fqn = build_fqn(context.module_path, None, node.name)

    item: DocItem = {
        "id": fqn,
        "name": node.name,
        "kind": ItemKind.FUNCTION,
        "language": Language.PYTHON,
        "location": get_location(node, file_path, source_dir),
        "metadata": {
            "signature": {
                "parameters": extract_parameters(node),
            },
        },
    }

    # Extract docstring
    docblock = extract_docblock(node)
    if docblock:
        item["docBlock"] = docblock

    # Extract return type annotation
    if node.returns:
        item["metadata"]["signature"]["returnType"] = {"name": ast.unparse(node.returns)}

    return item


def extract_parameters(node: ast.FunctionDef | ast.AsyncFunctionDef) -> List[Parameter]:
    """Extract parameters from a function."""
    parameters: List[Parameter] = []

    for arg in node.args.args:
        # Skip 'self' and 'cls'
        if arg.arg in ("self", "cls"):
            continue

        param: Parameter = {
            "name": arg.arg,
        }

        # Extract type annotation
        if arg.annotation:
            param["type"] = {"name": ast.unparse(arg.annotation)}

        parameters.append(param)

    return parameters


def extract_docblock(node: ast.AST) -> Optional[DocBlock]:
    """Extract docstring as a DocBlock."""
    docstring = ast.get_docstring(node)
    if not docstring:
        return None

    # Parse docstring
    parsed = parse_docstring(docstring)

    docblock: DocBlock = {}

    # Content
    if parsed.short_description or parsed.long_description:
        content_parts = []
        if parsed.short_description:
            content_parts.append(parsed.short_description)
        if parsed.long_description:
            content_parts.append(parsed.long_description)
        docblock["content"] = "\n\n".join(content_parts)

    # Tags - use Record<string, (string | DocTag)[]> format per spec
    tags: Dict[str, List[Any]] = {}

    # Parameters
    for param in parsed.params:
        tag: DocTag = {
            "name": TagName.PARAM,
            "content": param.description or "",
        }
        # Put parameter name and type in parameters object
        parameters: Dict[str, str] = {"name": param.arg_name}
        if param.type_name:
            parameters["type"] = param.type_name
        tag["parameters"] = parameters

        if TagName.PARAM not in tags:
            tags[TagName.PARAM] = []
        tags[TagName.PARAM].append(tag)

    # Returns
    if parsed.returns:
        return_content = parsed.returns.description or ""
        # Simple string for returns per spec examples
        if TagName.RETURNS not in tags:
            tags[TagName.RETURNS] = []
        tags[TagName.RETURNS].append(return_content)

    # Raises/Throws
    for raises in parsed.raises:
        tag: DocTag = {
            "name": TagName.THROWS,
            "content": raises.description or "",
            "parameters": {"name": raises.type_name}
        }
        if TagName.THROWS not in tags:
            tags[TagName.THROWS] = []
        tags[TagName.THROWS].append(tag)

    if tags:
        docblock["tags"] = tags

    return docblock if docblock else None


def get_location(node: ast.AST, file_path: Path, source_dir: Path) -> Location:
    """Get source location for a node."""
    return {
        "path": str(file_path.relative_to(source_dir)),
        "number": node.lineno,
        "column": node.col_offset + 1,
    }


def get_project_version(source_dir: Path) -> Optional[str]:
    """Get project version from pyproject.toml or setup.py."""
    # Try pyproject.toml
    pyproject_path = source_dir / "pyproject.toml"
    if pyproject_path.exists():
        try:
            if sys.version_info >= (3, 11):
                import tomllib
                with pyproject_path.open("rb") as f:
                    data = tomllib.load(f)
                    return data.get("project", {}).get("version")
            else:
                # For Python < 3.11, skip tomllib
                pass
        except Exception:
            pass

    # Try __version__ in __init__.py
    init_path = source_dir / "__init__.py"
    if init_path.exists():
        try:
            with init_path.open("r") as f:
                for line in f:
                    if line.startswith("__version__"):
                        return line.split("=")[1].strip().strip('"').strip("'")
        except Exception:
            pass

    return None
